auth/src/main/java/com/firebase/ui/auth/AuthUI.java

```diff
@@ -469,25 +546,265 @@ public Builder(@SupportedProvider @NonNull String providerId) {
+                GoogleSignInOptions.Builder builder =
+                        new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN);
+                for (String scope : scopes) {
+                    builder.requestScopes(new Scope(scope));
+                }
+                return setSignInOptions(builder.build());
+            }
+
+            /**
+             * Set the {@link GoogleSignInOptions} to be used for Google sign-in. Standard options
+             * like requesting the user's email will automatically be added.
+             *
+             * @param options sign-in options
+             */
+            @NonNull
+            public GoogleBuilder setSignInOptions(@NonNull GoogleSignInOptions options) {
+                if (getParams().containsKey(ExtraConstants.EXTRA_GOOGLE_SIGN_IN_OPTIONS)) {
+                    throw new IllegalStateException(
+                            "Cannot overwrite previously set sign-in options.");
+                }
+
+                GoogleSignInOptions.Builder builder = new GoogleSignInOptions.Builder(options);
+                builder.requestEmail().requestIdToken(getApplicationContext()
+                        .getString(R.string.default_web_client_id));
+                getParams().putParcelable(
+                        ExtraConstants.EXTRA_GOOGLE_SIGN_IN_OPTIONS, builder.build());
+
+                return this;
+            }
+
+            @NonNull
+            @Override
             public IdpConfig build() {
-                mScopes = mScopes == null ? Collections.<String>emptyList() : mScopes;
-                mParams = mParams == null ? new Bundle() : mParams;
+                if (!getParams().containsKey(ExtraConstants.EXTRA_GOOGLE_SIGN_IN_OPTIONS)) {
+                    setScopes(Collections.<String>emptyList());
+                }
+
+                return super.build();
+            }
+        }
+
+        /**
+         * {@link IdpConfig} builder for the Facebook provider.
+         */
+        public static final class FacebookBuilder extends Builder {
+            private static final String TAG = "FacebookBuilder";
+
+            public FacebookBuilder() {
+                //noinspection deprecation taking a hit for the backcompat team
+                super(FacebookAuthProvider.PROVIDER_ID);
+
+                try {
+                    //noinspection unused to possibly throw
+                    Class c = com.facebook.FacebookSdk.class;
+                } catch (NoClassDefFoundError e) {
+                    throw new RuntimeException(
+                            "Facebook provider cannot be configured " +
+                                    "without dependency. Did you forget to add " +
+                                    "'com.facebook.android:facebook-login:VERSION' dependency?");
+                }
+
+                if (getApplicationContext().getString(R.string.facebook_application_id)
```

Comment:
Same comment here, we should extract checking resources against the unconfigured value into a helper method.

Reply:
Done.